### ISSUE TITLE
[PROPOSAL] Author CHANGELOG in each PR instead of collecting them in the last days before a release

### ISSUE THREAD

**Issue by dblock** [opensearch-project/OpenSearch#1868]
**What kind of business use case are you trying to solve? What are your requirements?**

Release notes are of varying quality. 

For example, https://github.com/opensearch-project/OpenSearch/pull/2489 is completely unreadable.

**What is the problem? What is preventing you from meeting the requirements?**

- Release notes aren't always super useful and may lack context.
- Release notes are only available at release time, and not during development, so it's unclear what has changed.
- Each plugin team is responsible for release notes, and the person assembling the release notes may not be the best one to write about a specific change, so YMMV.

**What are you proposing? What do you suggest we do to solve the problem or improve the existing situation?**

1. Replace authoring release notes at the end before a release with authoring release notes in each PR.
2. Require that release notes be updated via something like [danger-changelog](https://github.com/dblock/danger-changelog) with every change.
3. Automate collection of release notes across all projects.
     https://github.com/opensearch-project/opensearch-build/issues/438
     https://github.com/opensearch-project/opensearch-build/issues/698

**Comment by stockholmux** [comment]
I see how this addresses `Release notes are only available at release time`, which I like, but I'm failing to understand how it address the other two points.

It seems like the output of [danger-changelog](https://github.com/dblock/danger-changelog/blob/master/CHANGELOG.md) is not too dissimilar to the [existing release notes](https://github.com/opensearch-project/opensearch-build/blob/main/release-notes/opensearch-release-notes-1.2.0.md) (last good example - subsequent 1.2.x are all log4j fixes), except without the context provided by the categories.

I will add that the release notes for a release do go through a human spot check. Often times that spot check will catch things like incorrect terminology, typos, or inconsistencies. Does the script driven approach allow for this?

**Comment by dblock** [comment]
> It seems like the output of [danger-changelog](https://github.com/dblock/danger-changelog/blob/master/CHANGELOG.md) is not too dissimilar to the [existing release notes](https://github.com/opensearch-project/opensearch-build/blob/main/release-notes/opensearch-release-notes-1.2.0.md) (last good example - subsequent 1.2.x are all log4j fixes), except without the context provided by the categories.

danger-changelog just checks that the PR includes a well-formatted entry for the changelog, it doesn't actually write it. The change I am proposing is that humans are required to author the changelog line and it gets reviewed as part of the PR
 
> I will add that the release notes for a release do go through a human spot check. Often times that spot check will catch things like incorrect terminology, typos, or inconsistencies. Does the script driven approach allow for this?

It doens't. I am proposing to remove any script-driven approach completely and move the authoring of release notes/changelog to the PRs, sorry if this wasn't clear.

**Comment by elfisher** [comment]
Would there be an expectation as part of the code review to review the release notes in the PR as well? That might help with quality.

**Comment by dblock** [comment]
> Would there be an expectation as part of the code review to review the release notes in the PR as well? That might help with quality.

Yes.

**Comment by peternied** [comment]
How about a format of the CHANGELOG that allows for cross checked against commits since last release?  This allows maintainers or contributors to add entries as needed or independently clean up the notes.  This avoids adding a barrier to contributions by adding another criteria on the PR to merge.

As part of health of the repo a tool can verify how many 'undocumented' items are not in the CHANGELOG.  This can be integrated into the release process during the post-code-freeze - prerelease time it can be required that maintainers drive that number to zero.  This allows for time to group similar changes and have dialogs about terminology/presentation.

**Comment by dblock** [comment]
@peternied Why would we need that if every PR _must_ have a CHANGELOG entry?

**Comment by peternied** [comment]
@dblock You are right, I am not thinking of chronological account of the changes in a codebase.

I want a reference of Most important -> Least important changes since last release.  Here is a contrived example of what I'd prefer vs changelog.   

---
### Release Clatu Verata Nictu
#### Breaking
* Removed indexing operators because they are not thread safe ([#1337](#))
#### Bug Fixes
* `Thread.current()` doesn't crash when called twice in a row ([#4567](#))
#### Features
* Added new lambda expressions, `map(...)`, `reduce(...)`, `sum(...)`. ([#44](#), [#45](#), [#46](#), [#4834](#))
---
vs 

### Release Clatu Verata Nictu
* Refactor sum to use reduce instead of its own implementation, `map(...)`, ([#4834](#))
* `Thread.current()` doesn't crash when called twice in a row ([#4567](#))
* Removed indexing operators because they are not thread safe ([#1337](#))
* Added new lambda expression, `reduce(...)`, ([#46](#))
* Added lambda expressions, `sum(...)`. ([#45](#))
* Added new lambda expressions, `map(...)`, ([#44](#))

**Comment by dblock** [comment]
@peternied We would adopt a standard, such as https://keepachangelog.com/en/1.0.0/

**Comment by peternied** [comment]
> @peternied We would adopt a standard, such as https://keepachangelog.com/en/1.0.0/

That standard format looks great, I am sold.

> Types of changes
> - Added for new features.
> - Changed for changes in existing functionality.
> - Deprecated for soon-to-be removed features.
> - Removed for now removed features.
> - Fixed for any bug fixes.
> - Security in case of vulnerabilities.

**Comment by getsaurabh02** [comment]
Thanks @dblock for the proposal. Another key area would be to standardize the various buckets or groups created for the release notes entries such as ``Enhancements``, ``Bug fixes``, ``Maintenance``, ``Refactoring``, ``Infrastructure`` etc. These are pretty insightful for readers to extract focussed information. Some of the recent release notes have these :

* https://github.com/opensearch-project/security/pull/1772/files
* https://github.com/opensearch-project/alerting/pull/426/files

Do we expect the ``CHANGELOG`` to have these groups pre-defined, that will allow authors/reviewers to add/review the CHANGELOG entry, added under the right group, along with every PR?

**Comment by dblock** [comment]
> Do we expect the `CHANGELOG` to have these groups pre-defined, that will allow authors/reviewers to add/review the CHANGELOG entry, added under the right group, along with every PR?

Yes, I expect those to be exactly from the list in https://keepachangelog.com/en/1.0.0/, at least to start.

**Comment by kartg** [comment]
> Require that release notes be updated via something like [danger-changelog](https://github.com/dblock/danger-changelog) with every change.

I'm guessing a "change" here could consist of multiple commits, as described by keepachangelog.org:

> The purpose of a commit is to document a step in the evolution of the source code. Some projects clean up commits, some don't.
The purpose of a changelog entry is to document the noteworthy difference, often across multiple commits, to communicate them clearly to end users.

In the case of multiple commits that gradually build up to a full feature, @dblock  is there a good mechanism we can put in place to make sure the changelog is reliably updated?

Alternatively, we could implements [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) as a precommit hook. That would give us a mechanism, but it still requires a changelog to be built and curated at release time. Thoughts?

**Comment by dblock** [comment]
@kartg I think we do simple: every PR requires a CHANGELOG update. Humans ensure quality in code reviews.

**Comment by kartg** [comment]
Works for me. In that case, we should define what makes a good Changelog entry - it shouldn't end up being just a repeat of the commit message.

**Comment by peternied** [comment]
@kartg @Poojita-Raj @dblock @getsaurabh02 @elfisher @stockholmux  I've opened https://github.com/opensearch-project/security/pull/1821 - would love your thoughts on how this looks

**Comment by rursprung** [comment]
> @kartg I think we do simple: every PR requires a CHANGELOG update. Humans ensure quality in code reviews. Multiple commits can be merged in CHANGELOG as we make progress.

from personal experience (we use this format in-house) i can tell you that, while this is a nice idea, it does result in a lot of merge conflicts (as they're all congesting on the same lines in `CHANGELOG.md`). i don't have a good solution for this though.

i do however 100% vote to follow [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) for the actual changelog! i've added some comments to the draft

**Comment by cliu123** [comment]
There've been the [release drafter](https://github.com/opensearch-project/security/blob/main/.github/workflows/release-drafter.yml) to generate [draft release notes](https://github.com/opensearch-project/security/releases/tag/untagged-7f7ae5bb3441abd9f61f) capturing all merged PRs. The release notes are generated by CI pipeline on each PR merge and avaialble for publishing releases. Do we still need this manually documented changelog?

**Comment by dblock** [comment]
@cliu123 I think the automation you propose produces complete unreadable garbage right now between backport PRs, DCO messages, etc. So it would solve having release notes continuously available, but it wouldn't help making these notes useful

**Comment by kotwanikunal** [comment]
After a dive into the discussion and release-drafter, it looks like release-drafter [does not directly support](https://github.com/release-drafter/release-drafter/issues/404#issue-556814039) multiple drafts on multiple branches. 
There is a [workaround available](https://github.com/release-drafter/release-drafter/issues/656#issuecomment-806184887) which can possibly support multiple draft notes. The caveat here is that we need a config for every version/branch in our case which is not scalable. 
We will have to either put in efforts to automate this using the script which could run with the version bump script but can still cause problems due to misconfigured `target_commitish` and the usage of GitHub UI for release publish. 

In the case we decide to fix all the issues with release-drafter or if we decide to move forward with a changelog based approach,  it would be a great addition to start labelling PRs to classify them into appropriate buckets. 
I have opened a PR for adding a label checker to PRs - #3618 

I am happy to chat and would love to move this discussion forward.

**Comment by dblock** [comment]
@kotwanikunal This proposal just says that the CHANGELOG would be edited by humans with every PR, thus you don't have to build any automation at all on branches to collect release notes, solves the problem you describe above

**Comment by kotwanikunal** [comment]
> @kotwanikunal This proposal just says that the CHANGELOG would be edited by humans with every PR, thus you don't have to build any automation at all on branches to collect release notes, solves the problem you describe above

>> @cliu123 I think the automation you propose would produce complete unreadable garbage right now between backport PRs, DCO messages, etc, at least on OpenSearch. So it would solve having release notes continuously available, but it wouldn't help making these notes useful. I could be wrong though, seems pretty cheap to try and enable it on this repo, want to try it? also how does it work across 3 ongoing releases in various branches?

I was referring to the mechanism used by plugins mentioned in the previous comment. The release-drafter approach does not make it any easier for us in terms of maintenance. 
Currently its difficult to compile these notes as they are and as a path forward, I was proposing labelling PRs to make it easier to classify even with the CHANGELOG. What do you think about that?

**Comment by dblock** [comment]
I don't see the point in labelling PRs if the said PR *must* also include a 1-line change in the appropriate section of the CHANGELOG in progress.

**Comment by kotwanikunal** [comment]
The proposal I had was something on the lines of -
- Add label checker for improving the current scenario (even manually generating the notes for that matter)
- Parallely/Later work on generating CHANGELOG for PRs in a format where the changelog generation itself is automated by a GHA which looks at the label and adds the changes (PR title/link) to the right location in the changelog 

I know the proposal originally required the requester to add the details to the CHANGELOG manually but I was trying to ease out that experience by going a bit further. 
With the current proposal, it will work more on the lines of an enforcer action than a generator. That should be an easy thing to add in if we are planning to go forward that

**Comment by Rishikesh1159** [comment]
Just wanted to document how I generated 2.1.0 release notes, so that this process might help in automation or for next release.
Steps:
```
* `git checkout origin/2.1`
* `git log --pretty=format:"* %s" --since=3-23-2022 > release-notes/opensearch.release-notes-2.1.0.md`
* Manually organize file sections
  * Add Date / Version section (copy + modify from previous release)
  * Enhancements/Bug fixes/Maintenance
* Replace pull request ids with full urls via regex
  * Find: `\(#(\d+)\)$`
  * Replace: `([#$1]([https://github.com/opensearch-project/opensearch/pull/$1))`](https://github.com/opensearch-project/opensearch/pull/$1))%60)
* Commit with these steps / modified with additional details
```
-> In second line of above command block, when using --since the date should be selected carefully. You should select the date when opensearch version bumped from MajorVersion.x to new version, example ([2.x version bump to 2.1](https://github.com/opensearch-project/OpenSearch/commit/f41d6a407103cd144509f5dda221affe4025d584))

-> I had to manually go through each commit and sort it out to one of the sections mentioned [here](https://github.com/opensearch-project/opensearch-plugins/blob/main/RELEASE_NOTES.md#categories).

-> Next I had to check which commits are already present in previous release version and remove them from current release notes manually. This step can be done before sorting out each commit to appropriate section.

->Make sure no breaking changes are included if it is a minor release versions.

**Comment by kotwanikunal** [comment]
@dblock I have worked on adding the workflow as a part of this PR - #4085 
The solution as it is currently implemented in the PR will not solve for backported PRs which will need additional work. (Maybe something on the lines of https://github.com/dangoslen/dependabot-changelog-helper) 
This however does solve for dependabot changes automatically and enforcing PRs for a manual changelog. 
Before we move ahead with the additional effort for backporting, I wanted to open this up for further discussion.

**Comment by kotwanikunal** [comment]
#4085 - change has been merged in and backported to `2.x` as well. 
Keeping the issue open to address any gaps that come up during the transition.

**Comment by kotwanikunal** [comment]
The changes have started showing results for the major unreleased branch and dependabot PRs. 
There are still some gaps around backporting that I have noticed and also called out as a part of #4085 - 
- There are frequent merge conflicts due to the sequence mismatch in backporting leading to a manual intervention
- There is no explicit entry for the change in the version to which the PR is backported to leading to an incomplete changelog for that version

A possible solution for the above problem is - 
- Drop the changelog file from https://github.com/VachaShah/backport/blob/main/src/backport.ts#L114-L119 and retry if the issue is with the CHANGELOG.md file
- Fork dangoslen/dependabot-changelog-helper plugin and adapt it for backport wherein we pull the original PR. locate the changed line for CHANGELOG.md, add an entry under the version for the backported branch  

Also, there is another discussion around missing process for feature branches here - #4477 

@dblock

**Comment by dblock** [comment]
Thanks, I think fixing backport to properly merge changelog is the way to go.

**Comment by kotwanikunal** [comment]
Raised a PR to add support for skipping possible conflicting files while backporting changes: https://github.com/VachaShah/backport/pull/4

**Comment by kotwanikunal** [comment]
Another update: I was able to fork https://github.com/dangoslen/dependabot-changelog-helper and add in additional features to backport the actual changelog line. 

It works exactly as I had described above -
- Takes in the original PR number as an input
- Uses the diff api from github to locate the change within `CHANGELOG.md`
- Extracts the changelog entry and section from the original PR
- Locates or creates the version entry within the CHANGELOG for the current backport branch, locates or creates the section within the version and adds the entry

Git: https://github.com/kotwanikunal/changelog-helper

I am waiting for the merge and release on https://github.com/VachaShah/backport/pull/4. We can then proceed with the calls for changelog helper.

**Comment by dblock** [comment]
@kotwanikunal is there a feature here you can try upstreaming into https://github.com/dangoslen/dependabot-changelog-helper? certainly worth your time I think in order to avoid maintaining a fork

**Comment by kotwanikunal** [comment]
> @kotwanikunal is there a feature here you can try upstreaming into https://github.com/dangoslen/dependabot-changelog-helper? certainly worth your time I think in order to avoid maintaining a fork

The current solution on my fork might be too specific to upstream. I'll try generalizing the approach and upstreaming it.

**Comment by kotwanikunal** [comment]
> IMO `files_to_skip` is a great feature for the upstream version. At the very least by creating a PR it create the opportunity for the maintainer to make a decision.

Thanks. I agree. That's the change that I have created for the `backport` GHA and I agree, that's upstreamable as is. I am just waiting to test it out.

I was discussing about an additional one that I have forked ([linked here](https://github.com/kotwanikunal/changelog-helper)) to enable line based cherry-picking of changelog using diffs from the original PR.

**Comment by rursprung** [comment]
something else i just noticed after contributing the first time since this rule has been established: it's rather impractical to add a reference to the PR in the changelog as that means that you first have to push your commit(s), create a (draft) PR, then update the changelog, commit & push that again (if you do it in the unclean way you just have an "update changelog" commit added, otherwise you need to briefly do some git-rebase voodoo to have a nice history). this adds an IMHO unnecessary step.

i think this could also be left away as the `CHANGELOG.md` is usually viewed somewhere where you can also run `git-blame` on it (GitHub's web-UI shows this; so does any modern IDE and on the console that of course works as well) and through that you can find the rest of the commit and from there the PR (when on GitHub).

WDYT?

**Comment by rursprung** [comment]
and one more thing: i just created https://github.com/opensearch-project/OpenSearch/pull/4723 which has no user-facing impact (purely internal to the build system, even w/o any impact there). IMHO this shouldn't be listed in the CHANGELOG at all (users don't care about this and reading about it just costs them time and confuses them until they find out that it has no impact for them). however, the current verification mechanism forces me to nevertheless list it there.

i'd suggest to relax this a bit (e.g. the ruling could be "all things relevant to users are documented in the `CHANGELOG.md`" and a reviewer can then always overrule the automated check?)

note: i write "user" but don't mean an end-user (who's calling the REST API or using OpenSearch Dashboards) but rather an operator of the software. maybe "operator" would be the better word? for commercial software i'd just write "customer", but obviously that wouldn't work here :)

**Comment by dblock** [comment]
> something else i just noticed after contributing the first time since this rule has been established: it's rather impractical to add a reference to the PR in the changelog as that means that you first have to push your commit(s), create a (draft) PR, then update the changelog, commit & push that again (if you do it in the unclean way you just have an "update changelog" commit added, otherwise you need to briefly do some git-rebase voodoo to have a nice history). this adds an IMHO unnecessary step.

I generally amend my commits. It's a small price to pay to help users locate the change by clicking on it IMO. 
 
> i think this could also be left away as the `CHANGELOG.md` is usually viewed somewhere where you can also run `git-blame` on it (GitHub's web-UI shows this; so does any modern IDE and on the console that of course works as well) and through that you can find the rest of the commit and from there the PR (when on GitHub).

Git blame is very technical and lacks a humanly readable explanation that is always in the body of the PR. So personally I am not for any changes from what we're trying to do now, even if it adds a tiny bit of work for every developer.

**Comment by kotwanikunal** [comment]
> i just noticed that the `CHANGELOG.md` of OpenSearch on `main` contains 3x the `Dependencies` heading for the `Unreleased` block. has this something to do with the automated dependabot functionality (i.e. a bug there) or has this just been missed by chance?
> 
> see here:
> 
> https://github.com/opensearch-project/OpenSearch/blob/77cff55bd431c0e2f8fdea5a40c83f8e8bddeca1/CHANGELOG.md?plain=1#L6
> 
> 
> https://github.com/opensearch-project/OpenSearch/blob/77cff55bd431c0e2f8fdea5a40c83f8e8bddeca1/CHANGELOG.md?plain=1#L36
> 
> 
> https://github.com/opensearch-project/OpenSearch/blob/77cff55bd431c0e2f8fdea5a40c83f8e8bddeca1/CHANGELOG.md?plain=1#L44

This seemed to be an issue because we were using spaces along with heading within CHANGELOG.md which was creating new entries for the unreleased version. I have fixed that with #4740

**Comment by dblock** [comment]
FYI, see https://github.com/opensearch-project/OpenSearch/issues/4769 for an ask to revert all this. Please add your comments.

**Comment by seraphjiang** [comment]
> 1) users/contributors didn't know what changed half way during development

could we use release issue to track to avoid merge issue?

> 2) the data was wrong too often

we don't have to collect all commit, we could define the rule for commit message/label 

> 3) other information that's very hard to read for users.

again, we should improve our commit history, insist highest bar on it. not ignore it


> 4) 1) is a small amount of work and a developer can be expected to have to describe their change in a reasonable 1-liner.

this is not truh

> 1. when many develop work on same project, they have to rebase multiple times if they didn't get their PR merged after they got 1) Approval, 2) rebase because other change the changelog 3) passed tests . any of condition happen cause developer to redo 1) rebase fix change log 2) run tests, 3) collect approval again.  

> 2) is being worked on in backport automation.

this is not true, what I heard and observe from OSD cases, 90% backport case has conflict.


my asking is 
could we invest on automation release note other than introduce this manual process? I'm also not asking for investing on automation changelog

could we use meta release issue to track change, that way we don't introduce a global lock file `changelog`

**Comment by kotwanikunal** [comment]
Reading through #4769 and this, here are my thoughts and what is currently baking - 

1. On PR conflicts:
- Agreeably, there have been conflicts for `CHANGELOG.md` for PRs since we started with this effort
- Within the `OpenSearch` repository, do not need re-approvals in case of new commits on a PR, hence the effort is not as extensive with fixing the changelog

2. On backporting: 
- I am looking at updating our current workflows with backporting the literal changelog line to the needed version of the backported branch.
- This will eliminate all the conflict related issues and backporting related issues that have been called out
- More on that here: https://github.com/opensearch-project/OpenSearch/issues/1868#issuecomment-1258827491

> could we invest on automation release note other than introduce this manual process? 
I would love to have some alternative to achieve this and can put in the effort if you have suggestions around such a framework. I agree with @dblock that, as a release manager, collecting notes is painful and developer/contributor notes are the best approach to a human-readable changelog. I am also all ears on alternatives that folks have to suggest.

**Comment by seraphjiang** [comment]
I propose using change log issue to replace existing change log file approach.

Keep the same format as changelog file.

When PR got merged, maintainer and author will decide if it is critical change need to add to change log.


When backport to release branch, it should update item in changelog issue to right version section

With this approach we still has place to take release note but without worry merge conflict hell anymore.

Any concern on this approach

**Comment by anasalkouz** [comment]
Here is my thought regarding the changelog file:
I do agree to keep the changelog file approach and try to avoid any human interaction during the process and I don't expect merging a conflict in the changelog file should take more than few mins to resolve. That being said, I think we need to invest more time to make it more easier for developers and remove an churn by 1) doing some automation improvements like what @kotwanikunal suggested to reduce number of conflicts. and 2) Make merging changes to the changelog optional and it's up to the author and reviewers to decide, this will help incase we have PRs with feature to announce like code refactor adding unit test ...etc.

**Comment by andrross** [comment]
I think the current process of requiring _every_ commit to produce a line in the changelog is a big part of the problem here. We are essentially just duplicating the one-liner header from the commit message into the changelog, filed under the appropriate category. If this is the desired goal, then I think it can be achieved by requiring discipline around the git commit message along with git tooling (this is how the linux kernal does it, for example).

@seraphjiang proposed solution suggests this is _not_ the goal, as he states "When PR got merged, maintainer and author will decide if it is critical change need to add to change log." Getting clarity on this should help us find a solution. @dblock What is the ideal solution here? Should there be a 1-to-1 relationship between git commits and changelog entries, or should this be a manually curated list of changes?

**Comment by Bukhtawar** [comment]
I think its also something that we can templatise or follow best practises https://github.com/torvalds/subsurface-for-dirk/blob/a48494d2fbed58c751e9b7e8fbff88582f9b2d02/README#L88 while writing a commit message. The maintainer can then use his/her discretion to choose what qualifies for a change log and appropriately ensure it gets published

**Comment by seraphjiang** [comment]
Thanks @anasalkouz  @andrross and @Bukhtawar @dblock  for discussion


It is true, change log conflict take mins to resolve, however it then it introduce new commit, and need to re-run CI which takes hours to finish. If this happen again and again when multiple people is committing, developer will be very frustrated. 

Think about long term productivity when more and more contribute join the effort (they don't even has permission to re-run failed flaky test). The changelog file process increases the their time and cost to contribute which leads to frustration. 

Because majority of contributors don't has permission to backlog, maintainer will have to resolve merge conflict for the backport change(which is not related to actual code itself)

Thinking out of box, the goal is to address release note issue, changelog if just one option. We knew we need **1) automation**  and **2) best practices and discipline, template** to enforce good commit message, no matter we has changelog or not. 

I'll try an experiment for [release note automation](https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes#configuring-automatically-generated-release-notes) and share what I learn from it.

**Comment by seraphjiang** [comment]
I have tested [release note automation](https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes#configuring-automatically-generated-release-notes)

it is pretty neat, see my experiment below 

https://github.com/seraphjiang/dashboards-anywhere/releases/tag/v0.9.1

**Comment by dblock** [comment]
@seraphjiang I like automation. Try this on a fork of OpenSearch, let's see  how readable the result is? Enable it on main, 2.x and a released version, e.g. 1.3.5. Also can those automated release notes be edited post-fact or will they forever contain typos that developers make in commit messages?

**Comment by dblock** [comment]
@andrross I think we want a 1:1 relationship, but we also want the changes to be humanly readable

**Comment by dblock** [comment]
> Thinking out of box, the goal is to address release note issue, changelog if just one option. We knew we need **1) automation** and **2) best practices and discipline, template** to enforce good commit message, no matter we has changelog or not.

Re: 2, it's asking a lot from code reviewers to check commit messages, and for contributors to amend and force-push their changes again and again

**Comment by seraphjiang** [comment]
> @seraphjiang I like automation. Try this on a fork of OpenSearch (this repo), let's see how readable the result is? Enable it on main, 2.x and a released version, e.g. 1.3.5. Also can those automated release notes be edited post-fact or will they forever contain typos that developers make in commit messages? And what happens when versions change branches or two branches have the same version (happens all the time, right now plugins are struggling to increment version on main to 3.0)?

tried clone the opensearch repo, unfortunately, it will reset the PR history not clone that as well. 

https://github.com/seraphjiang/OpenSearch/releases

we has add release.yml in existing repo, will have a try and update here.
https://github.com/opensearch-project/dashboards-anywhere

**Comment by dblock** [comment]
Is this how it would look? https://github.com/seraphjiang/OpenSearch/releases/tag/2.2.0

**Comment by dblock** [comment]
Throwing in https://www.conventionalcommits.org/en/v1.0.0-beta.2/ which could become a standard way to write commit messages. This is used in https://github.com/aws/sagemaker-python-sdk and https://github.com/aws/copilot-cli.

**Comment by dblock** [comment]
@kotwanikunal I am seeing that we have both Unreleased and 2.x sections in main and 2.x. Was that intentional? Should everything be unreleased?

**Comment by seraphjiang** [comment]
> Is this how it would look? https://github.com/seraphjiang/OpenSearch/releases/tag/2.2.0

not exactly, but similar, the format is configuable.

**Comment by kotwanikunal** [comment]
> @kotwanikunal I am seeing that we have both Unreleased and 2.x sections in main and 2.x. Was that intentional? Should everything be unreleased? where do backports go?

It was supposed to be Unreleased for main, 2.x for 2.x. Backporting has led to a mess and inconsistency, with folks not adding the change to 2.x.
Currently `Unreleased` is the only one which has all the correct changes, and I was hoping to get main + backporting in line since that will bring the other branches to a correct state automatically with the helper.

**Comment by seraphjiang** [comment]
@dblock this is how it looks without adding labels like feature/bug/enhancement/break

https://github.com/opensearch-project/dashboards-anywhere/releases/tag/v0.8.0

here is release.yml
https://github.com/opensearch-project/dashboards-anywhere/blob/main/.github/release.yml

**Comment by seraphjiang** [comment]
another example

https://github.com/opensearch-project/opensearch-dashboards-functional-test/releases/tag/v2.4.0-rc

**Comment by seraphjiang** [comment]
this is github's public rest api to generate release note, which we could integrate with our release automation

https://docs.github.com/en/rest/releases/releases#generate-release-notes-content-for-a-release

**Comment by rursprung** [comment]
> > @kotwanikunal I am seeing that we have both Unreleased and 2.x sections in main and 2.x. Was that intentional? Should everything be unreleased? where do backports go?
> 
> It was supposed to be Unreleased for main, 2.x for 2.x. Backporting has led to a mess and inconsistency, with folks not adding the change to 2.x. Currently `Unreleased` is the only one which has all the correct changes, and I was hoping to get main + backporting in line since that will bring the other branches to a correct state automatically with the helper.

tbh, that sounds wrong. the changelog for 2.x should only live on the 2.x branch - the main branch should only have the changelog for main (i.e. the moment 2.x is branched away the two changelogs start to diverge). otherwise you have a nightmare in maintaining the changelog (as you have now).

i know that i was the one who advocated for the currently used changelog and still think that the _format_ is correct. i however don't think that the way it's currently being done is good (esp. after i've now hassled with it a few times when creating PRs):
* i think it's wrong to require an entry for _every_ commit - as i've said from the beginning, the changelog should be for the _users_ of the software; they don't care about reformatting the code, changing test coverage, etc.
* creating the entries with every PR leads to a ton of merge conflicts (in smaller projects this generally works fine as they have less PRs landing all the time, here it's just a merging/rebasing nightmare) - though i think if only the relevant things are added then this will also be less of an issue as there'd be much fewer entries

right now i've nearly never managed to get in a PR directly: either it failed with a false reject (lots of flaky tests around; not the subject of this issue here) and/or it ran into a merge conflict due to the changelog by the time it got an approval (or the tests were re-run). since i don't work every day (and sign all my commits with a PGP key which is of course different for work & private) the PR then has to sit around for a while (usually a week) until i get around to rebasing it again, which is a huge loss of time and might mean that it doesn't land in a specific release in the worst case.

**Comment by rursprung** [comment]
just as a minor clarification: when i wrote "for every commit" i actually meant "for every PR". but that doesn't change a lot as larger features still consist of multiple PRs (leading to multiple entries for one new feature instead of a single entry) and there are lots of PRs which are not relevant for users/operators.

**Comment by dblock** [comment]
@rursprung @andrross I have a `skip-changelog` label in https://github.com/dblock/create-a-github-issue that could be implemented here for dependabot and backports, see https://github.com/dblock/create-a-github-issue/pull/16/files. WDYT? Anyone wants to make that change here and cleanup the CHANGELOG in main and 2.x?

**Comment by dblock** [comment]
@seraphjiang Want to try to PR what you're suggesting into OpenSearch main?

**Comment by andrross** [comment]
From [keepachangelog.com](https://keepachangelog.com/en/1.0.0/#bad-practices):

> Can changelogs be bad?
> 
> Yes. Here are a few ways they can be less than useful.
> Commit log diffs
> Using commit log diffs as changelogs is a bad idea: they're full of noise. Things like merge commits, commits with obscure titles, documentation changes, etc.

I think our 1:1 requirement that all PRs have a changelog entry results in us implementing this bad practice. The changelog ends up being slightly more readable than a commit log (entries are categorized) but we have no way to filter out the commits that are not useful to users. Ultimately "useful to users" is a judgment call, so I think we need a process here that allows for the developer and maintainer to make a judgment about whether to include a changelog entry for a given PR.

**Comment by ashwin-pc** [comment]
Having read through as much of this thread as possible, I dont think i've seen anyone suggest this, so apologies if this was already discussed. Cant we decouple the change-log edits from the PR's responsible for the change (with some automation to make sure that tracking the two becomes easier).

1. We add an automated label `changelog` to every PR that needs a changelog entry (new entry or update). It automatically creates a new issue in the repo thats assigned to the author of the PR. 
2. The new Issue also has a link to the PR, A summary of the change and other goodies to make creating a Changelog change as easy and traceable as possible.
3. If the PR has any backport tags, the changelog issue will also have those tags called out for tracking

Pros:
- Decouples fixes and changes from changelog (Less chances for merge conflicts)
- Changelog is more intentional
- Goes better with the dev flow of raising a PR and adding a changelog afterwards for the change (No need to force push a change to update just the changelog)
- Changelog PR's are quick followups since all the tests need not run for those PR's (Markdown only PR's are wayyy quicker than other PR's)

Cons: 
- Changes and Changelog entries are in 2 separate PR's. This is mitigated by having references in both the Original PR and Changelog issue to each other.

**Comment by ashwin-pc** [comment]
Another suggestion i want to drop in here is GitLabs approach that they documented in this blog post: https://about.gitlab.com/blog/2018/07/03/solving-gitlabs-changelog-conflict-crisis/

**Comment by andrross** [comment]
Here's the PR (#5067) for creating the release notes for 2.4 in OpenSearch. I manually merged entries into a single "feature" entry where appropriate, which is desirable in my opinion (breaking the 1-entry-for-every-PR constraint currently enforced). We had some duplicate/incorrect entries split between the [Unreleased] and [2.x] section of the file. Overall it wasn't too bad to clean up.

That PR was on the 2.4 branch. I believe the CHANGELOG on the 2.x branch can be emptied out now (all changes from 2.x were released into 2.4). The main branch is a mess though. It contains all changes that have been released, interleaved with changes that have not been backported and will remain unreleased until 3.0. This process is not working well with our branch strategy, specifically the fact that we're committing next-major-version changes alongside next-minor-version changes that get backported. All next-minor-version changes should not have an entry on the main branch because they will never be "unreleased" from the perspective of the next-major-version.

**Comment by dblock** [comment]
@andrross Yes, I agree it's not working :( Let's keep looking at @kotwanikunal to figure out best path forward until he tells us otherwise?

**Comment by andrross** [comment]
@dblock @kotwanikunal 

I've written up an FAQ to be included in the CHANGELOG.md file in #5092 

This defines a slight change in the process. First, it specifies not every PR should include a changelog entry (and assumes we'll have a `skip-changelog` label. Second, it defines how the changelog should work with our branching strategy. Please take a look.

**Comment by andrross** [comment]
Just an update after the 2.5 release...unfortunately the latest process is also not working great. The problems seemed to mostly arise around [where should I put my CHANGELOG entry](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#where-should-i-put-my-changelog-entry). In an ideal world, the `[Unreleased 2.x]` section on the main branch should be identical to the `[Unreleased 2.x]` section on the 2.x branch. The process to create release notes should be: move the `[Unreleased 2.x]` section on main to a `release-notes-2.5.0.md` file and then backport to the 2.x and 2.5 branches. In practice that did not work because some changes on main were put into the `[Unreleased 3.0]` section but then backported and put into the `[Unreleased 2.x]` section on the 2.x branch (this is easy for reviewers to miss because it is unintuitive you have to expand out the CHANGELOG file to even see what section the entry is in). @kotwanikunal ultimately ended up scanning through commits and manually auditing to create the release notes file.

**Comment by nknize** [comment]
I mentioned this in a [PR comment](https://github.com/opensearch-project/OpenSearch/pull/6009#issuecomment-1404237990) and will post it here since I haven't followed this moving target closely.

I suggest we switch the changelog to adopt a more standard approach like other projects and use the issue number instead of the PR. It would enforce having an issue for each PR, and stop this burden of pushing a changelog commit that updates the url in each PR.

**Comment by dblock** [comment]
@nknize Not everything has an issue, but the additional commit is actually a relatively small problem compared to the fact that CHANGELOG in every PR creates a ton of additional work and backport merge conflicts, even if done right. It solved one problem, but created a different one. 

Another idea is to adopt a different mechanism for PR titles, and settle on something that generates CHANGELOGs in every branch automatically from PR titles that can be changed after merge. Check out https://github.com/aws/aws-cdk/pulls for what that could look like.

**Comment by nknize** [comment]
> Not everything has an issue

If it doesn't have an issue it doesn't need to be in the changelog.

**Comment by andrross** [comment]
> > Not everything has an issue
> 
> If it doesn't have an issue it doesn't need to be in the changelog.

I'm in favor of linking issues instead of PRs, particularly since we updated that guidance that [not all changes require a CHANGELOG entry](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#which-changes-require-a-changelog-entry) and added the `skip-changelog` label.

> It solved one problem, but created a different one.

Honestly it didn't really even solve a problem (at least for the OpenSearch repo) for this latest release as the CHANGELOG was a bit of a mess between the two branches and required a lot of manual reconciliation to create the release notes. I've been nagging folks to make sure entries go in the right place since that release but I don't think we've arrived at a sustainable solution.
